# **Exploratory Data Analysis (EDA) for Mars Weather Analysis**

## **1. Introduction**
### **Dataset Overview**
- The dataset consists of Martian weather data collected by the **REMS instrument on NASA's Curiosity rover**.
- It contains **3,197 records** spanning multiple Martian years, with variables covering **temperature, pressure, UV radiation, and day length**.

### **Objectives of EDA**
- Understand feature relationships and trends.
- Identify missing data patterns and their impact.
- Investigate seasonal and cyclical trends.
- Detect outliers and anomalies.

### **Key Questions to Explore**
- Are temperature and pressure correlated?
- Can features predict missing values?
- How do seasonal and yearly trends affect Martian weather?
- What role does **day length** play in temperature variations?
- Are there **outliers or extreme events** in the dataset?

### EDA TO DO LIST
- [ ] Investigate each feature in my dataframe
- [ ] Pearson correlation coefficients plots
- [ ] Hypothesis testing and establishment of probing questions
1. Are temperature and pressure correlated?
2. Can any of the features be used to predict the gaps in the data?
- [ ] Identify if there are any years with complete sol records, this will be the primary training dataset
- [ ] Look at seasonal trends for primary features
- [ ] Look into seasonal changes in variation between min and max temperatures
- [ ] Look into distributions of missing days and data (from original dataset)

---

## **2. Data Overview**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

### **2.1 Load and Inspect Data**
- Load the dataset (`pd.read_csv()`).
- Display first few rows (`df.head()`).
- Check data structure (`df.info()`) and column data types.
- Convert date columns to datetime format (`pd.to_datetime()`).

In [8]:
mars_df = pd.read_csv('../Datasets/mars_weather_cleaned.csv', index_col='sol_number')
mars_df.head()

earth_date_time     Ls  mars_month  mars_year mars_season  \
sol_number                                                             
1               2012-08-07  150.0           6          1      winter   
9               2012-08-15  155.0           6          1      winter   
10              2012-08-16  155.0           6          1      winter   
11              2012-08-17  156.0           6          1      winter   
12              2012-08-18  156.0           6          1      winter   

             sunrise    sunset  day_length  avg_ground_temp  max_ground_temp  \
sol_number                                                                     
1           05:30:00  17:22:00       712.0            -45.5            -16.0   
9           05:28:00  17:22:00       714.0            -45.5            -16.0   
10          05:28:00  17:22:00       714.0            -45.5            -16.0   
11          05:28:00  17:21:00       713.0            -43.5            -11.0   
12          05:28:00  17:21:00       713.0            -47.0            -18.0   

            min_ground_temp  avg_air_temp  max_air_temp  min_air_temp  \
sol_number                                                              
1                     -75.0         -37.5           8.0         -83.0   
9                     -75.0         -37.5           8.0         -83.0   
10                    -75.0         -37.5           8.0         -83.0   
11                    -76.0         -37.0           9.0         -83.0   
12                    -76.0         -37.0           8.0         -82.0   

            mean_pressure  UV_Radiation  
sol_number                               
1                   739.0             4  
9                   739.0             4  
10                  739.0             4  
11                  740.0             4  
12                  741.0             4

### **2.2 Summary Statistics**
- Generate summary statistics (`df.describe()`).
- Identify potential outliers and missing values.

In [9]:
mars_df.describe()

Ls   mars_month    mars_year   day_length  avg_ground_temp  \
count  3197.000000  3197.000000  3197.000000  3197.000000      3197.000000   
mean    166.959962     6.083203     3.507038   718.169221       -44.097279   
std     104.356771     3.466172     1.500218    12.036363         7.401663   
min       0.000000     1.000000     1.000000   702.000000       -72.500000   
25%      78.000000     3.000000     2.000000   707.000000       -50.500000   
50%     156.000000     6.000000     3.000000   716.000000       -43.000000   
75%     254.000000     9.000000     5.000000   730.000000       -37.500000   
max     359.000000    12.000000     6.000000   738.000000       -26.500000   

       max_ground_temp  min_ground_temp  avg_air_temp  max_air_temp  \
count      3197.000000      3197.000000   3197.000000   3197.000000   
mean        -13.182828       -75.011730    -39.147357      2.011730   
std          10.489177         5.529929      7.262277      9.398862   
min         -67.000000      -100.000000    -75.500000    -61.000000   
25%         -23.000000       -79.000000    -45.500000     -6.000000   
50%         -12.000000       -75.000000    -38.000000      3.000000   
75%          -4.000000       -71.000000    -33.000000     10.000000   
max          11.000000       -52.000000     -1.500000     24.000000   

       min_air_temp  mean_pressure  UV_Radiation  
count   3197.000000    3197.000000   3197.000000  
mean     -80.306537     828.997028      2.583359  
std        8.824723      57.224328      0.692132  
min     -136.000000     702.000000      1.000000  
25%      -86.000000     785.000000      2.000000  
50%      -80.000000     844.000000      3.000000  
75%      -75.000000     873.000000      3.000000  
max       -8.000000     925.000000      4.000000

### **2.3 Missing Data Analysis**
- **Count missing values** per column (`df.isnull().sum()`).
- **Visualize missing data**:
  - `msno.matrix(df)` to see patterns in missing values.
  - `msno.heatmap(df)` to explore correlations between missing features.
- **Identify patterns in missing data**:
  - Check if missing data is seasonal (`sns.heatmap(df.pivot_table(index='mars_month', columns='mars_year', values='avg_air_temp', aggfunc='count'))`).

---

## **3. Feature Investigation**

### **3.1 Univariate Analysis (Distributions & Trends)**
- **Histograms** (`sns.histplot()`) for:
  - `avg_air_temp`, `max_air_temp`, `min_air_temp`
  - `avg_ground_temp`, `max_ground_temp`, `min_ground_temp`
  - `mean_pressure`, `UV_Radiation`, `day_length`
- **Boxplots** (`sns.boxplot()`) to detect outliers.

### **3.2 Bivariate Analysis (Feature Relationships)**
- **Scatter plots** (`sns.scatterplot()`):
  - `avg_air_temp` vs. `mean_pressure`
  - `avg_ground_temp` vs. `UV_Radiation`
  - `day_length` vs. `max_air_temp`
- **Boxplots by categorical variables**:
  - Seasonal temperature variation: `sns.boxplot(x='mars_season', y='avg_air_temp')`
  - Monthly pressure changes: `sns.boxplot(x='mars_month', y='mean_pressure')`

---

## **4. Correlation Analysis**

### **4.1 Pearson Correlation**
- Compute **correlation matrix** (`df.corr()`).
- **Heatmap of correlations** (`sns.heatmap(df.corr(), annot=True, cmap='coolwarm')`).
- Investigate relationships:
  - `avg_air_temp` vs. `mean_pressure`
  - `max_air_temp` vs. `UV_Radiation`
  - `day_length` vs. `temperature`

In [10]:
mars_df.corr()

ValueError: could not convert string to float: '2012-08-07'

### **4.2 Pairwise Comparisons**
- **Pair plot visualization** (`sns.pairplot(df[['avg_air_temp', 'mean_pressure', 'UV_Radiation']])`).
- Check for **nonlinear relationships** using **Spearman correlation**.

---

## **5. Missing Data Analysis**

### **5.1 Investigating Missing Patterns**
- **Time-series visualization** of missing values:
  - `df.groupby('mars_year').apply(lambda x: x.isnull().sum())`
  - Plot missing values over time using `sns.heatmap()`.

### **5.2 Predicting Missing Data**
- **Regression-based approaches**:
  - Use **Linear Regression or KNN Imputation** to check if missing values can be estimated using other variables.
- **Hypothesis Testing**:
  - Compare distributions of missing vs. non-missing data (`sns.boxplot(x='Missing', y='avg_air_temp')`).
  - Perform **t-tests** (`scipy.stats.ttest_ind()`) to check if missing data follows systematic trends.

---

## **6. Cyclical Trends: Seasonal and Yearly Patterns**

### **6.1 Identify Years with Complete Sol Records**
- **Check completeness**:  
  - `df.groupby('mars_year')['earth_date_time'].count()`
  - Filter for years with **full sol records**.

### **6.2 Seasonal Variability**
- **Line plots** of monthly trends:
  - `sns.lineplot(x='mars_month', y='avg_air_temp', hue='mars_year', data=df)`
- **Compare seasons**:
  - `sns.boxplot(x='mars_season', y='avg_air_temp', data=df)`

---

## **7. Day Length vs. Temperature Relationship**
- **Scatter Plots**:
  - `sns.scatterplot(x='day_length', y='max_ground_temp', data=df)`
- **Correlation Test** (`scipy.stats.pearsonr(df['day_length'], df['max_air_temp'])`).

---

## **8. UV Radiation and Atmospheric Conditions**
- **UV vs. Temperature**: `sns.scatterplot(x='UV_Radiation', y='avg_air_temp')`.
- **Seasonal UV Levels**: `sns.boxplot(x='mars_season', y='UV_Radiation')`.
- **Correlation Analysis**:
  - `df[['UV_Radiation', 'avg_air_temp', 'mean_pressure']].corr()`

---

## **9. Differences Between Air & Ground Temperatures**
- Compute **temperature gap**:
  - `df['temp_gap'] = df['avg_ground_temp'] - df['avg_air_temp']`
- **Visualize over time**: `sns.lineplot(x='mars_month', y='temp_gap', data=df)`.
- **Seasonal Variability**:
  - Boxplots for temp gaps across seasons.

---

## **10. Long-Term Trends Over Martian Years**
- **Rolling averages** to smooth trends:
  - `df['mean_pressure'].rolling(window=30).mean().plot()`
- **Compare yearly distributions**:
  - `sns.boxplot(x='mars_year', y='mean_pressure', data=df)`

---

## **11. Outlier and Anomaly Detection**
- **Boxplots for outliers**: `sns.boxplot(x=df['mean_pressure'])`.
- **Z-Score Analysis**:
  - `df['z_score'] = (df['mean_pressure'] - df['mean_pressure'].mean()) / df['mean_pressure'].std()`
  - Identify extreme z-scores (`df[df['z_score'] > 3]`).

---

## **12. Hypothesis Testing**

### **12.1 Are Temperature and Pressure Correlated?**
- **Null Hypothesis (H₀):** No correlation.
- **Alternative Hypothesis (H₁):** There is a correlation.
- **Statistical Test**:
  - Pearson correlation test (`scipy.stats.pearsonr()`).

### **12.2 Can Features Predict Missing Data?**
- Compare distributions using **t-tests**.
- Explore regression-based imputation.

---

## **13. Summary of Findings & Next Steps**
- **Key insights**: Seasonal trends, missing data impact, correlations, anomalies.
- **Potential next steps**:
  - Feature engineering.
  - Model selection for missing data prediction.
  - Time-series forecasting models.